### ISSUE TITLE
Fix events management

### ISSUE THREAD

**Issue by alexhrao** [alexhrao/TheGiftHub#51]
# Problem

There is ZERO support for adding, editing, or deleting events. This requires forethought - we have to make sure the current database setup is ideal before we begin anything else!

# Steps

## 1. Requirements

What, exactly, are we trying to do here? How will we do it? We need to sort this out before any code is written!

## 2. Database

We need to make sure that the database is setup nicely for our requirements

## 3. Back End

We’ll need to rewrite event handling code - especially handling event futures!

## 4. Front end

We also need to handle how the user will interact with these events. They’ll need to be able to:

- Create an event
- Delete an event
- Share the event with other groups
- Pick a default event
- Add event futures
- Use a default event as a base for further edits
- Define an infinitely recurring event if it’s on the same day every year
- Define custom future dates
- **Future plans** Allow the user to add “custom rules”

**Comment by alexhrao** [comment]
Progress has been made; below are details of this progress:

**TL;DR**: **_The "Future Plans" are now!_**
# Progress

## Requirements

This part is finished, below are the results

A user should be able to:
- Create an event that only happens once (a *"one-off"*)
- Create an event that occurs on some "exact" time scale. Examples include:
  - Every year starting on March 1st
  - Every 1st of the month
  - Every other Tuesday
  - Every 3 days
- Create an event that occurs on some "relative" time scale. Examples include:
  - Every year on the third Thursday of March
  - The first Monday of every month
  - Every other month, on the last Friday of that month, ending on `SomeDate`
- For any recurring event, have "blackout dates", which are days when there is an exception to the rule and the event should *not* occur.

### Exact Events

Exact Events are events that occur on an "exact" time scale. This does **not** mean that the time itself never varies; indeed, with Daylight Savings Time and Leap Year, it undoubtedly will. Instead, Exact means that the event begins on it's `StartDate` and happens every `TimeInterval`.

As an example, an Exact Event could be defined as starting on January 10th, 2017 and happening every month. In this case, the next event occurrences would be:
- February 10th, 2017
- March 10th, 2017
- April 10th, 2017

This doesn't depend on the "week" it's in, or on a specific day of the week.

### Relative Events

Relative events are events whose time interval is relative to something else - there is no "exact" time span. So, if a user wanted an event to occur every year on the third Thursday of January, this isn't possible to do with Exact Events. Instead, we have Relative Events.

As an example, a Relative Event could be defined as starting on January 7th, 2017, and occurring every year on the second Monday of January, and should stop by January 1st, 2020. In this case, the next event occurrences would be:
- January 9th, 2017
- January 8th, 2018
- January 14th, 2019

Note that the start date doesn't necessarily indicate the first day the event occurs - it just specifies a lower bound. Furthermore, the end date doesn't necessarily indicate the last occurrence - only an upper bound.

### Blackout Dates

Blackout dates can be applied to any event, and define a specific date where, although the rules would imply an event occurrence, there should not be one. This is an exceptional event.

As an example, suppose the user wanted an event to occur every 5th of the month, starting on January 5th, 2017, except for in February 2017. In this case, the next event occurrences would be:
- March 5th, 2017
- April 5th, 2017
- May 5th, 2017

**Note that at this time, there is no way to say "Every month except January". Blackout dates are on a single date-by-date basis.**

## Database Design

This part is also finished. Below are the results.

The design consists of four tables:
- `user_events`
- `exact_events`
- `relative_events`
- `blackout_events`

### Events Table

The `user_events` table has the following fields:
- `EventID`             **PK**
  - Integer, Auto Increment, Unique & Non-Null
  - The Primary Key for this `event`
- `UserID`               **FK**
  - Integer, Non-Null
  - The `user` who owns this `event`
- `EventName`
  - Variable Character, 256 Character limit & Non-Null
  - The name of this event
- `EventStartDate`
  - Date, Non-Null
  - The first occurrence of this event, or at least the lower bound for event occurrences
- `EventEndDate`
  - Date, Null
  - The last occurrence of this event, or at least the upper bound for event occurrences
  - *NB: If this is `NULL`, then the event repeats in perpetuity*

### Exact Events Table

The `exact_events` table has the following fields:
- `ExactEventID`     **PK**
  - Integer, Auto Increment, Unique & Non-Null
  - The Primary Key for this `exact event` rule
- `EventID`              **FK**
  - Integer, Unique & Non-Null
  - The `event` this rule applies to
- `EventTimeInterval`
  - Single Character & Non-Null
  - Can only be one of the following values:
    - 'D'   (Daily)
    - 'W'  (Weekly)
    - 'M'  (Monthly)
    - 'Y'    (Yearly)
  - The interval at which this `event` recurs.
- `EventSkipEvery`
  - Integer, Positive & Non-Null
  - Describes the recurrence pattern - 1 is "every", 2 is "every other", 3 is "every third", etc.

### Relative Events Table

The `relative_events` table has the following fields:
- `RelativeEventID` **PK**
  - Integer, Auto Increment, Unique & Non-Null
  - The Primary Key for this `relative event` rule
- `EventID`              **FK**
  - Integer, Unique & Non-Null
  - The `event` this rule applies to
- `EventTimeInterval`
  - Single Character & Non-Null
  - Can only be one of the following values:
    - 'M' (Monthly)
    - 'Y' (Yearly)
  - The interval at which this `event` recurs
- `EventSkipEvery`
  - Integer, Positive & Non-Null
  - Describes the recurrence pattern - 1 is "every", 2 is "every other", 3 is "every third", etc.
- `EventDayOfWeek`
  - Single Character & Non-Null
  - Can only be one of the following values:
   - 'N' (Sunday)
   - 'M' (Monday)
   - 'T' (Tuesday)
   - 'W' (Wednesday)
   - 'R' (Thursday)
   - 'F' (Friday)
   - 'S' (Saturday)
  - Which day of the week this `event` recurs with
- `EventPosn`
  - Integer, Positive & Non-Null
  - Can only be one of the following values:
   - 1 (The first `DayOfWeek`)
   - 2 (The second `DayOfWeek`)
   - 3 (The third `DayOfWeek`)
   - 4 (The fourth `DayOfWeek`)
   - 5 (The last `DayOfWeek`)
  - The position within the month this `event` recurs

### Blackout Events Table

The `blackout_events` table has the following fields:
- `BlackoutEventID` **PK**
  - Integer, Auto Increment, Unique & Non-Null
  - The Primary Key for this `blackout event`
- `EventID` **FK**
  - Integer, Unique & Non-Null
  - The `event` this blacks out
- `EventBlackoutDate`
  - Date, Non-Null
  - The date this `event` should not occur on

## Back End

This consists of implementing the following classes:

- ### `Event`
  - #### Implements:
    - `ISynchronizable`
    - `IFetchable`
  - #### Fields:
    - ##### `EventId`
      - `ulong`
      - The ID for this `Event`
    - ##### `Name`
      - `string`
      - The Name for this `Event`
    - ##### `StartDate`
      - `DateTime`
      - The date this `Event` starts on (or at least a lower bound)
    - ##### `EndDate`
      - `DateTime`
      - The date this `Event` stops recurring (or at least the upper bound)
      - **Note: If this is `null`, then the event never stops recurring!**
    - ##### `User`
      - `User`
      - The `User` that owns this `Event`
    - ##### `Rules`
      - `RulesEngine`
      - The rules that dictate how this `Event` recurs
      - **Note: If this is `null`, then the event does not recur**
    - ##### `Blackouts`
      - `List<Blackout>`
      - A list of all blackout dates for this event
      - Cannot be `null` but *can* be an empty list
  - #### Methods:
    - ##### `bool Create()` *From `ISynchronizable`*
      - Creates a record of this `Event` in the database
    - ##### `bool Update()` *From `ISynchronizable`*
      - Updates an existing record for this `Event` in the database
    - ##### `bool Delete()` *From `ISynchronizable`*
      - Delete an existing record for this `Event` from the database
    - ##### `XmlDocument Fetch()` *From `IFetchable`*
      - Serializes this `Event` as an XML Document
  - #### Notes:
    - This is the main class for events.
- ### `RulesEngine`
  - #### Implements:
    - `ISynchronizable`
    - `IFetchable`
  - #### Fields:
    - ##### `Event`
      - The `Event` this rule applies to.
  - #### Methods:
    - ##### `abstract IEnumerator<DateTime> Occurrences;`
      - Implemented as an `abstract property`
      - Gets all occurrences of this event, starting with the first occurrence
      - **Note: This has the potential to run infinitely, if the event never stops recurring. It is up to the caller to decide when enough events have passed**
    - ##### `abstract bool Create()` *From `ISynchronizable`*
      - Creates a record of this `Rule` in the database
    - ##### `abstract bool Update()` *From `ISynchronizable`*
      - Updates an existing record for this `Rule` in the database
    - ##### `abstract bool Delete()` *From `ISynchronizable`*
      - Delete an existing record for this `Rule` from the database
    - ##### `abstract XmlDocument Fetch()` *From `IFetchable`*
      - Serializes this `Rule` as an XML Document
  - #### Notes:
    - This class is `abstract`, and must be implemented by either `ExactEvent` or `RelativeEvent`
- ### `ExactEvent`
  - #### Derives:
    - `RulesEngine`
  - #### Fields:
    - ##### `ExactEventId`
      - `ulong`
      - The ID for this `ExactEvent`
    - ##### `TimeInterval`
      - `string`
      - ###### Can only be one of following values:
        - 'D' (Daily)
        - 'W' (Weekly)
        - 'M' (Monthly)
        - 'Y' (Yearly)
      - The interval that separates event occurrences
    - ##### `SkipEvery`
      - `uint`
      - Must be greater than 0
      - How many intervals to skip before the next occurrence
  - #### Methods:
    - *Implements all abstract methods from `RulesEngine`*
  - #### Notes:
    - This is a rule for an event that occurs on an exact time table (*i.e., every year, every other week, etc.*)
- ### `RelativeEvent`
  - #### Derives:
    - `RulesEngine`
  - #### Fields:
    - ##### `RelativeEventId`
      - `ulong`
      - The ID for this `RelativeEvent`
    - ##### `TimeInterval`
      - `string`
      - ###### Can only be one of following values:
        - 'M' (Monthly)
        - 'Y' (Yearly)
      - The interval that separates event occurrences
    - ##### `SkipEvery`
      - `uint`
      - Must be greater than 0
      - How many intervals to skip before the next occurrence
    - ##### `DayOfWeek`
      - `string`
      - ###### Can only be one of the following values:
        - 'N' (Sunday)
        - 'M' (Monday)
        - 'T' (Tuesday)
        - 'W' (Wednesday)
        - 'R' (Thursday)
        - 'F' (Friday)
        - 'S' (Saturday)
      - The day of the week this `Event` occurs on
    - ##### `Posn`
      - `uint`
      - ###### Can only be one of the following values:
        - `1` (First of...)
        - `2` (Second of...)
        - `3` (Third of...)
        - `4` (Fourth of...)
        - `5` (Last of...)
      - The position within the month this `Event` occurs
  - #### Methods:
    - *Implements all abstract methods from `RulesEngine`*
  - #### Notes:
    - This is a rule for an event that occurs on a relative timetable (*i.e., The third Thursday of every month, etc.*)
- ### `Blackout`
  - #### Implements:
    - `ISynchronizable`
    - `IFetchable`
  - #### Fields:
    - ##### `BlackoutId`
      - `ulong`
      - The ID for this `Blackout`
    - ##### `Event`
      - `Event`
      - The `Event` for which this `Blackout` applies
    - ##### `BlackoutDate`
      - `DateTime`
      - The Date for which this `Event` should not occur, regardless of rules
  - #### Methods:
    - ##### `bool Create()` *From `ISynchronizable`*
      - Creates a record of this `Blackout` in the database
    - ##### `bool Update()` *From `ISynchronizable`*
      - Updates an existing record for this `Blackout` in the database
    - ##### `bool Delete()` *From `ISynchronizable`*
      - Delete an existing record for this `Blackout` from the database
    - ##### `XmlDocument Fetch()` *From `IFetchable`*
      - Serializes this `Blackout` as an XML Document
  - #### Notes:
    - A single `Blackout` is really just a date where the event should not occur.

# Next Steps

This is wonderful progress. Below are specific items that must be completed for this issue to close

## Database

- [x] Delete the "old" event management tables
- [x] Create the "new" event management tables
- [x] Edit the "load data" script to account for this new way of managing events

## Back End

- [x] Delete the "old" classes that corresponded to event management
- [x] Implement the "new" classes we have documented here
- [x] Integrate these classes into the Controller

## Front end

- [x] Add the ability to create a new event from the profile page
- [x] Add the ability to delete an existing event from the profile page
- [ ] Add the ability to edit an existing event from the profile page

Now that we have the Requirements, Database Design, and Class Design complete, we can begin implementing this! All work will be done on the [AddEvents](https://github.com/alexhrao/TheGiftHub/tree/AddEvents) Branch.

**Comment by alexhrao** [comment]
Arguably, we don't need to be able to "edit" events - simply delete the event and set it up again!

Other than this, this feature is finished development - now it has to be tested

**Comment by alexhrao** [comment]
Minor problem with a relative event - if the user selects monthly, no problem, we can store it. However, if the user only wants it yearly, then we have no way of storing the month.

# Solution

Add a field called EventMonth to RelativeEvents - this will allow us to store that month there, and it can be null so that if we are doing it monthly, we don't need it.

## Proposed Changes

- ### Database
  - `relative_events`
    - Change the `Interval` field:
      - `CHAR(3)`
      - Nullable
      - Represents three-letter code for the month, if it occurs yearly. Otherwise, it's null
- ### Back End
  - `RelativeEvent`
    - Change the `TimeInterval` field accordingly - make it a string instead!
